File: runtime/db-postgres.js
// =============================================================================
// CLEAR RUNTIME — DATABASE MODULE (PostgreSQL backend)
// =============================================================================
//
// Same API as db.js (SQLite) — drop-in replacement for cloud deployments.
// Uses pg.Pool with DATABASE_URL from environment.
//
// API (matches db.js exactly):
//   db.createTable(name, schema)  — registers schema, lazy-creates on first query
//   db.findAll(table, filter?, options?) — SELECT * with optional WHERE + LIMIT
//   db.findOne(table, filter)     — SELECT * WHERE ... LIMIT 1
//   db.aggregate(table, fn, field, filter?) — SELECT FN(col) ... with equality filter
//   db.insert(table, record)      — INSERT ... RETURNING *
//   db.update(table, filter, data?)— UPDATE matching records
//   db.remove(table, filter?)     — DELETE matching records
//   db.run(sql)                   — execute raw SQL
//   db.execute(sql)               — alias for db.run
//   db.save()                     — no-op (Postgres is always durable)
//   db.load()                     — no-op
//   db.reset()                    — TRUNCATE all known tables
//
// Key difference from db.js: all query functions are async.
// Compiled Clear code already uses `await` on all db calls, so this is safe.
// createTable() is synchronous (stores schema only) because it's called at
// module load time before any await is possible. Tables are created lazily
// on the first actual query via ensureTable().
//
// =============================================================================

'use strict';

const { Pool } = require('pg');
const { AsyncLocalStorage } = require('async_hooks');

// Defer DATABASE_URL check to first query — don't crash on require() for health-check-only servers
var pool = null;
function getPool() {
  if (pool) return pool;
  if (!process.env.DATABASE_URL) {
    throw new Error('[clear:db] DATABASE_URL not set. Add a Postgres database in your Railway dashboard.');
  }
  pool = new Pool({
    connectionString: process.env.DATABASE_URL,
    // SSL for cloud Postgres (Railway, Render, etc.)
    ssl: process.env.DATABASE_URL.includes('sslmode=') ? undefined : { rejectUnauthorized: true },
  });
  process.on('SIGTERM', function() { pool.end(); });
  return pool;
}

// Schema registry — populated synchronously by createTable(), used by lazy init
const _schemas = {};
// Track which tables have been created in Postgres (avoids re-running CREATE TABLE)
const _tablesCreated = new Set();
// Track which tables have had RLS enabled (avoids re-running policy DDL)
const _rlsEnabled = new Set();
const IDENT_RE = /^[a-z_][a-z0-9_]*$/i;

// =============================================================================
// TENANT-SCOPED ROW-LEVEL SECURITY (defense in depth)
// =============================================================================
//
// When the source declares `database is shared with tenant scope`, the
// compiled app already injects `tenant_id = req.user.tenant_id` into every
// app-layer CRUD filter. RLS adds a second layer: the database itself
// refuses cross-tenant reads/writes, even if a future bug or raw-SQL slip
// bypasses the app filter.
//
// How it works:
//   1. Express middleware wraps each authenticated request with
//      `withTenantScope(req.user.tenant_id, next)`. AsyncLocalStorage
//      threads the tenant id through every nested await.
//   2. Every CRUD call in the request goes through `_query`, which detects
//      the tenant context, opens a transaction, runs `SET LOCAL
//      app.current_tenant_id = <id>`, runs the actual query, and commits.
//      SET LOCAL clears at COMMIT/ROLLBACK, so the pooled connection is
//      safe to reuse for the next request without a stale var.
//   3. `enableRowLevelSecurity(table)` (called once per shared-scope table
//      at app startup) runs `ALTER TABLE ... ENABLE ROW LEVEL SECURITY`
//      and creates a policy that requires `tenant_id =
//      current_setting('app.current_tenant_id')::int` for every row read,
//      written, updated, or deleted.
//
// Net effect: Postgres physically refuses to return another tenant's rows.
// A future query that forgets the WHERE clause still reads zero foreign
// rows because the policy filters them out at the database layer.
//
// =============================================================================

const _tenantStore = new AsyncLocalStorage();

// Public: runs `fn` inside an async-local tenant context. Every CRUD call
// nested inside `fn` (no matter how deep the await chain) automatically
// scopes its query to this tenant via `SET LOCAL app.current_tenant_id`.
// Pass `null`/`undefined` for no scope (CRUD goes through pool directly).
function withTenantScope(tenantId, fn) {
  if (tenantId === undefined || tenantId === null) return fn();
  var n = parseInt(tenantId, 10);
  if (!Number.isFinite(n)) return fn();
  return _tenantStore.run({ tenantId: n }, fn);
}

// Internal: returns the current tenant id, or null when no scope is active.
function _currentTenantId() {
  var ctx = _tenantStore.getStore();
  return ctx && Number.isFinite(ctx.tenantId) ? ctx.tenantId : null;
}

// Internal: runs a SQL query. When a tenant scope is active, wraps the
// query in BEGIN + SET LOCAL + COMMIT so the database-level RLS policy
// fires. When no scope is active, goes straight to the pool (current
// behavior preserved for non-shared-scope apps and pre-auth requests).
async function _query(sql, params) {
  var tenantId = _currentTenantId();
  if (tenantId === null) {
    return await getPool().query(sql, params || []);
  }
  var client = await getPool().connect();
  try {
    await client.query('BEGIN');
    // tenantId is already a finite integer (validated in withTenantScope),
    // so direct interpolation is safe — no SQL injection surface.
    await client.query('SET LOCAL app.current_tenant_id = ' + tenantId);
    var res = await client.query(sql, params || []);
    await client.query('COMMIT');
    return res;
  } catch (e) {
    try { await client.query('ROLLBACK'); } catch (_) {}
    throw e;
  } finally {
    client.release();
  }
}

// Public: enables row-level security on a table and creates the
// tenant-isolation policy. Idempotent — safe to call repeatedly. Bypasses
// the tenant context (uses the pool directly) because policy DDL must run
// as the table owner, not as a tenant. Called once per shared-scope table
// at app startup, after createTable.
async function enableRowLevelSecurity(tableName) {
  if (!IDENT_RE.test(tableName)) throw new Error('Invalid table name: ' + tableName);
  var lc = tableName.toLowerCase();
  if (_rlsEnabled.has(lc)) return;
  // Make sure the table exists before trying to enable RLS on it.
  await ensureTable(lc);
  // ENABLE RLS — turn on policy enforcement.
  await getPool().query('ALTER TABLE ' + lc + ' ENABLE ROW LEVEL SECURITY');
  // FORCE RLS — apply policies even to the table owner (without this, the
  // owner connection bypasses RLS, which defeats defense-in-depth: a bug
  // running as the owner could still leak cross-tenant rows).
  await getPool().query('ALTER TABLE ' + lc + ' FORCE ROW LEVEL SECURITY');
  // Drop-and-recreate so policy text changes propagate on redeploy.
  // Postgres 16+ supports CREATE POLICY IF NOT EXISTS; we use the older
  // pattern for compat with 12/13/14/15.
  await getPool().query('DROP POLICY IF EXISTS clear_tenant_isolation ON ' + lc);
  await getPool().query(
    'CREATE POLICY clear_tenant_isolation ON ' + lc +
    ' FOR ALL TO PUBLIC' +
    " USING (tenant_id = current_setting('app.current_tenant_id', true)::int)" +
    " WITH CHECK (tenant_id = current_setting('app.current_tenant_id', true)::int)"
  );
  _rlsEnabled.add(lc);
}

// =============================================================================
// TYPE HELPERS
// =============================================================================

function toPgType(config) {
  if (!config || !config.type) return 'TEXT';
  switch (config.type) {
    case 'number': return 'DOUBLE PRECISION';
    case 'boolean': return 'BOOLEAN';
    case 'fk': return 'INTEGER';
    case 'timestamp': return 'TIMESTAMPTZ';
    default: return 'TEXT';
  }
}

function toPgDefault(config) {
  if (config.default === undefined) return '';
  if (config.type === 'boolean') return ' DEFAULT ' + (config.default ? 'TRUE' : 'FALSE');
  if (config.type === 'number') return ' DEFAULT ' + config.default;
  return " DEFAULT '" + String(config.default).replace(/'/g, "''") + "'";
}

function stripHidden(row, schema) {
  if (!row || !schema) return row;
  var out = Object.assign({}, row);
  for (var field in schema) {
    if (!schema.hasOwnProperty(field)) continue;
    if (schema[field] && schema[field].hidden) delete out[field];
  }
  return out;
}

async function backfillRenamedFields(tableName, schema, existing) {
  if (!IDENT_RE.test(tableName) || !schema) return;
  for (var fromField in schema) {
    if (!schema.hasOwnProperty(fromField)) continue;
    var config = schema[fromField];
    var toField = config && config.renamedTo;
    if (!config || !config.hidden || !toField) continue;
    if (!IDENT_RE.test(fromField) || !IDENT_RE.test(toField)) continue;
    if (!existing.has(fromField) || !existing.has(toField)) continue;
    await getPool().query(
      'UPDATE ' + tableName +
      ' SET "' + toField + '" = "' + fromField + '"' +
      ' WHERE "' + toField + '" IS NULL AND "' + fromField + '" IS NOT NULL'
    );
  }
}

// =============================================================================
// VALIDATION (same as db.js — application-level constraints before SQL)
// =============================================================================

// Strip dangerous HTML patterns to prevent stored XSS
function sanitizeRecord(record) {
  if (!record || typeof record !== 'object') return record;
  var result = {};
  for (var key in record) {
    if (!record.hasOwnProperty(key)) continue;
    var value = record[key];
    if (typeof value === 'string') {
      result[key] = value
        .replace(/<script\b[^>]*>[\s\S]*?<\/script>/gi, '')
        .replace(/on\w+\s*=\s*["'][^"']*["']/gi, '')
        .replace(/on\w+\s*=\s*[^\s>]*/gi, '')
        .replace(/javascript\s*:/gi, '');
    } else {
      result[key] = value;
    }
  }
  return result;
}

// Coerce string values to their declared types (number, boolean)
function enforceTypes(record, schema) {
  if (!schema) return;
  for (var field in schema) {
    if (!schema.hasOwnProperty(field)) continue;
    var config = schema[field];
    if (record[field] === undefined || record[field] === null) continue;
    if (config.type === 'number') {
      // Empty string -> null, not 0. Number("") === 0 is silent data corruption.
      if (record[field] === '' || (typeof record[field] === 'string' && record[field].trim() === '')) {
        record[field] = null;
        continue;
      }
      var num = Number(record[field]);
      if (isNaN(num)) throw new Error(field + ' must be a number, got ' + JSON.stringify(record[field]));
      record[field] = num;
    }
    if (config.type === 'boolean' && typeof record[field] === 'string') {
      if (record[field] === 'true') record[field] = true;
      else if (record[field] === 'false') record[field] = false;
    }
  }
}

// Fill in default values for fields not provided
function applyDefaults(record, schema) {
  if (!schema) return record;
  var result = Object.assign({}, record);
  for (var field in schema) {
    if (!schema.hasOwnProperty(field)) continue;
    var config = schema[field];
    if (result[field] === undefined && config.default !== undefined) result[field] = config.default;
    if (result[field] === undefined && config.auto && config.type === 'timestamp') {
      result[field] = new Date().toISOString();
    }
  }
  return result;
}

// Check that all required fields are present and non-empty
function validateRequired(record, schema) {
  if (!schema) return null;
  for (var field in schema) {
    if (!schema.hasOwnProperty(field)) continue;
    var config = schema[field];
    if (config.required && (record[field] === undefined || record[field] === null || record[field] === '')) {
      return field + ' is required';
    }
  }
  return null;
}

// =============================================================================
// LAZY TABLE CREATION
// =============================================================================

// Creates the table in Postgres on first use. Called before every query.
// createTable() just stores the schema; this does the actual DDL.
async function ensureTable(tableName) {
  if (_tablesCreated.has(tableName)) return;
  var schema = _schemas[tableName];
  if (!schema) { _tablesCreated.add(tableName); return; }

  // Build CREATE TABLE with all columns from schema
  var cols = ['id SERIAL PRIMARY KEY'];
  for (var field in schema) {
    if (!schema.hasOwnProperty(field)) continue;
    // Quote column names to handle Postgres reserved words (e.g. "order", "user")
    cols.push('"' + field + '" ' + toPgType(schema[field]) + toPgDefault(schema[field]));
  }
  await getPool().query('CREATE TABLE IF NOT EXISTS ' + tableName + ' (' + cols.join(', ') + ')');

  // Schema evolution: add columns that exist in schema but not in table
  var res = await getPool().query(
    "SELECT column_name FROM information_schema.columns WHERE table_name = $1",
    [tableName]
  );
  var existing = new Set(res.rows.map(function(r) { return r.column_name; }));
  for (var field2 in schema) {
    if (!schema.hasOwnProperty(field2)) continue;
    if (!existing.has(field2)) {
      await getPool().query('ALTER TABLE ' + tableName + ' ADD COLUMN IF NOT EXISTS "' + field2 + '" ' + toPgType(schema[field2]) + toPgDefault(schema[field2]));
      existing.add(field2);
    }
  }
  await backfillRenamedFields(tableName, schema, existing);
  _tablesCreated.add(tableName);
}

// =============================================================================
// TABLE MANAGEMENT
// =============================================================================

// Synchronous — just registers the schema. Called at module load time.
// Actual table creation happens lazily in ensureTable() on first query.
// Table name is sanitized to prevent SQL injection from malicious Clear source.
function createTable(name, schema) {
  var tableName = name.toLowerCase().replace(/[^a-z0-9_]/g, '');
  if (!tableName) throw new Error('Invalid table name: ' + name);
  _schemas[tableName] = schema || {};
}

// =============================================================================
// FILTER -> SQL WHERE
// =============================================================================

// Builds a parameterized WHERE clause from a filter object.
// Returns { clause, params, offset } where offset is the count of params used
// (so subsequent SET clauses can continue numbering from $offset+1).
function buildWhere(filter) {
  if (!filter || Object.keys(filter).length === 0) return { clause: '', params: [], offset: 0 };
  var conditions = [];
  var params = [];
  var i = 1;
  for (var key in filter) {
    if (!filter.hasOwnProperty(key)) continue;
    // Validate column name to prevent SQL injection via filter keys
    if (!/^[a-zA-Z_][a-zA-Z0-9_]*$/.test(key)) continue;
    conditions.push('"' + key + '" = $' + i);
    params.push(filter[key]);
    i++;
  }
  return { clause: 'WHERE ' + conditions.join(' AND '), params: params, offset: i - 1 };
}

// =============================================================================
// CRUD OPERATIONS
// =============================================================================

function parseLimit(n) {
  var v = parseInt(n, 10);
  return (v > 0 && v < 10000) ? v : null;
}

function parseOffset(n) {
  var v = parseInt(n, 10);
  return (v >= 0 && v < 1000000) ? v : null;
}

async function findAll(table, filter, options) {
  var tableName = table.toLowerCase();
  await ensureTable(tableName);
  var schema = _schemas[tableName] || {};
  var includeHidden = !!(options && options.includeHidden);
  var w = buildWhere(filter);
  var sql = 'SELECT * FROM ' + tableName + ' ' + w.clause;
  if (options && options.limit) {
    var lim = parseLimit(options.limit);
    if (lim) sql += ' LIMIT ' + lim;
  }
  if (options && options.offset) {
    var off = parseOffset(options.offset);
    if (off) sql += ' OFFSET ' + off;
  }
  var res = await _query(sql, w.params);
  return includeHidden ? res.rows : res.rows.map(function(row) { return stripHidden(row, schema); });
}

async function findOne(table, filter, options) {
  var tableName = table.toLowerCase();
  await ensureTable(tableName);
  var schema = _schemas[tableName] || {};
  var includeHidden = !!(options && options.includeHidden);
  var w = buildWhere(filter);
  var res = await _query('SELECT * FROM ' + tableName + ' ' + w.clause + ' LIMIT 1', w.params);
  if (!res.rows[0]) return null;
  return includeHidden ? res.rows[0] : stripHidden(res.rows[0], schema);
}

function validateAggregateArgs(fn, field) {
  var allowedFns = { SUM: 1, AVG: 1, MIN: 1, MAX: 1, COUNT: 1 };
  if (!allowedFns[fn]) throw new Error('Unsupported aggregate function: ' + fn);
  if (fn !== 'COUNT' && !/^[a-z_][a-z0-9_]*$/i.test(field)) {
    throw new Error('Invalid field name: ' + field);
  }
}

async function aggregate(table, fn, field, filter) {
  var tableName = table.toLowerCase();
  await ensureTable(tableName);
  validateAggregateArgs(fn, field);
  var w = buildWhere(filter);
  var col = fn === 'COUNT' ? '*' : field;
  var sql = 'SELECT ' + fn + '(' + col + ') as result FROM ' + tableName + ' ' + w.clause;
  try {
    var res = await _query(sql, w.params);
    return res.rows[0] ? (res.rows[0].result || 0) : 0;
  } catch (e) {
    console.warn('[clear] db.aggregate failed:', e.message);
    return 0;
  }
}

async function insert(table, record) {
  var tableName = table.toLowerCase();
  var schema = _schemas[tableName] || {};
  await ensureTable(tableName);

  record = sanitizeRecord(record);
  enforceTypes(record, schema);

  var reqErr = validateRequired(record, schema);
  if (reqErr) throw new Error(reqErr);

  // Unique constraint check (application-level, before SQL).
  // Under tenant scope, uniqueness becomes per-tenant — different tenants
  // can have the same email/slug because RLS hides the other tenant's rows.
  // That's the intended semantics: tenant A's "deal-001" doesn't block
  // tenant B from creating their own "deal-001".
  for (var field in schema) {
    if (!schema.hasOwnProperty(field)) continue;
    if (!schema[field].unique || record[field] === undefined) continue;
    var check = await _query(
      'SELECT 1 FROM ' + tableName + ' WHERE "' + field + '" = $1 LIMIT 1', [record[field]]
    );
    if (check.rows.length > 0) throw new Error(field + " must be unique -- '" + record[field] + "' already exists");
  }

  // Foreign key check (application-level). Under tenant scope, FK targets
  // are also tenant-filtered — you can't reference another tenant's row
  // because RLS won't let the SELECT see it.
  for (var fkField in schema) {
    if (!schema.hasOwnProperty(fkField)) continue;
    if (schema[fkField].type !== 'fk') continue;
    var value = record[fkField];
    if (value === undefined || value === null || value === '') continue;
    var refTable = schema[fkField].ref ? schema[fkField].ref.toLowerCase() : (fkField.endsWith('_id') ? fkField.replace(/_id$/, '') + 's' : null);
    if (!refTable) continue;
    if (!refTable.endsWith('s')) refTable += 's';
    await ensureTable(refTable);
    var fkCheck = await _query('SELECT 1 FROM ' + refTable + ' WHERE id = $1 LIMIT 1', [value]);
    if (fkCheck.rows.length === 0) throw new Error(fkField + ' references non-existent record (id ' + value + ' not found in ' + refTable + ')');
  }

  var withDefaults = applyDefaults(record, schema);
  var fields = Object.keys(withDefaults).filter(function(k) { return k !== 'id'; });

  if (fields.length === 0) {
    var res = await _query('INSERT INTO ' + tableName + ' DEFAULT VALUES RETURNING *');
    return res.rows[0];
  }

  var placeholders = fields.map(function(_, idx) { return '$' + (idx + 1); });
  var values = fields.map(function(f) { return withDefaults[f]; });
  var quotedFields = fields.map(function(f) { return '"' + f + '"'; });
  var res2 = await _query(
    'INSERT INTO ' + tableName + ' (' + quotedFields.join(', ') + ') VALUES (' + placeholders.join(', ') + ') RETURNING *',
    values
  );
  return res2.rows[0];
}

async function update(table, filterOrRecord, data) {
  var tableName = table.toLowerCase();
  var schema = _schemas[tableName] || {};
  await ensureTable(tableName);

  var filter, updateData;
  if (data === undefined) {
    // Convention 1: db.update('table', record) — update by record.id
    var record = filterOrRecord;
    if (record.id !== undefined) {
      filter = { id: record.id };
      updateData = record;
    } else {
      return 0;
    }
  } else {
    // Convention 2: db.update('table', filter, data)
    filter = filterOrRecord;
    updateData = data;
  }

  updateData = sanitizeRecord(updateData);
  enforceTypes(updateData, schema);

  var w = buildWhere(filter);
  if (!w.clause) return 0;

  // Guard: throw 404 when updating by id but record doesn't exist.
  // Under tenant scope, "exists" also means "exists in this tenant" — a
  // cross-tenant id will 404 because RLS hides the row.
  if (filter.id !== undefined) {
    var exists = await _query('SELECT 1 FROM ' + tableName + ' ' + w.clause + ' LIMIT 1', w.params);
    if (exists.rows.length === 0) {
      var err = new Error('No record found with id ' + filter.id);
      err.status = 404;
      throw err;
    }
  }

  var setCols = Object.keys(updateData).filter(function(k) { return k !== 'id'; });
  if (setCols.length === 0) return 0;

  // SET params continue numbering after WHERE params
  var setEntries = setCols.map(function(k, i) { return '"' + k + '" = $' + (w.offset + i + 1); });
  var setValues = setCols.map(function(k) { return updateData[k]; });

  var sql = 'UPDATE ' + tableName + ' SET ' + setEntries.join(', ') + ' ' + w.clause;
  var result = await _query(sql, w.params.concat(setValues));
  return result.rowCount;
}

async function remove(table, filter) {
  var tableName = table.toLowerCase();
  await ensureTable(tableName);
  var w = buildWhere(filter);
  var result = await _query('DELETE FROM ' + tableName + ' ' + w.clause, w.params);
  return result.rowCount;
}

// =============================================================================
// RAW SQL
// =============================================================================

async function run(sql) {
  await _query(sql);
}

async function execute(sql) {
  return run(sql);
}

// =============================================================================
// LIFECYCLE
// =============================================================================

function save() { /* no-op: Postgres is always durable */ }
function load() { /* no-op */ }

async function reset() {
  for (var tableName in _schemas) {
    if (!_schemas.hasOwnProperty(tableName)) continue;
    try {
      await getPool().query('TRUNCATE TABLE ' + tableName + ' RESTART IDENTITY CASCADE');
    } catch (e) {
      // Table might not exist yet — ignore
    }
  }
}

// =============================================================================
// PUBLIC API
// =============================================================================

module.exports = {
  createTable: createTable,
  findAll: findAll,
  findOne: findOne,
  insert: insert,
  update: update,
  remove: remove,
  aggregate: aggregate,
  run: run,
  execute: execute,
  save: save,
  load: load,
  reset: reset,
  // Tenant-scoped row-level security (defense in depth)
  withTenantScope: withTenantScope,
  enableRowLevelSecurity: enableRowLevelSecurity,
};

File: runtime/db_postgres.py
"""
CLEAR RUNTIME — POSTGRES DATABASE MODULE (Python port)

Python port of runtime/db-postgres.js. Drop-in replacement for runtime/db.py
when the compiled Clear app is deployed against a real PostgreSQL database
(Railway, Render, Fly Postgres, RDS, etc.) instead of local SQLite.

API matches runtime/db.py exactly (snake_case for PEP 8). Same names,
same call shapes — the compiled Python emit picks one or the other based
on the source's `database is postgres` vs default declaration.

Library dep: `psycopg[binary]>=3.0` (PyPI).
    pip install 'psycopg[binary]>=3.0'

INTEROP: the schema this module creates matches what runtime/db-postgres.js
creates on the same DATABASE_URL — same column names, same types, same
auto-added (id PRIMARY KEY, user_id, tenant_id, _version) columns. A row
inserted by the JS Postgres runtime reads back via this Python module
(and vice versa). Use case: a Clear app deployed with the JS target on
Railway can be re-deployed with the Python target on the same DB without
data migration.

DEFERRED to follow-up (matches db.py's pattern):

- **`update_with_version(table, where, data)`** — optimistic-lock UPDATE.
  Stubbed with NotImplementedError + plan pointer, same as in db.py.
- **Tenant-scoped Row-Level Security context (`AsyncLocalStorage` in JS,
  `contextvars` in Python).** The JS port uses AsyncLocalStorage to pass
  `tenant_id` through every nested call so RLS can enforce isolation.
  Python equivalent: `contextvars.ContextVar`. Stubbed for v1; the
  underlying `SET LOCAL app.current_tenant_id` SQL runs on every query
  but the value is currently a no-op default.
- **Sensitive-field encrypt-at-rest integration with sensitive_crypto.py.**
  db.py has the lazy-import + per-call encrypt/decrypt; this module marks
  the integration points but doesn't wire them yet.

These deferred pieces are tracked in plans/plan-python-parity.md.
Today's scaffold gets Python apps to the point where they can persist
data on Postgres; the optimistic-lock + tenant-RLS-context + sensitive-
field integrations follow once the basic CRUD is verified end-to-end
against a real Postgres instance.
"""

import os
import re
import threading
from typing import Any, Dict, List, Optional, Union

try:
    import psycopg  # type: ignore
    from psycopg.rows import dict_row  # type: ignore
    _PSYCOPG_AVAILABLE = True
except ImportError:
    _PSYCOPG_AVAILABLE = False

# Defer DATABASE_URL check to first query — don't crash on import for
# health-check-only servers (matches db-postgres.js behavior).
_pool_lock = threading.Lock()
_pool: Optional[Any] = None  # psycopg.ConnectionPool when initialized


def _get_pool():
    global _pool
    if _pool is not None:
        return _pool
    with _pool_lock:
        if _pool is not None:
            return _pool
        if not _PSYCOPG_AVAILABLE:
            raise RuntimeError(
                "[clear:db_postgres] psycopg is not installed. "
                "Run: pip install 'psycopg[binary]>=3.0'"
            )
        if not os.environ.get("DATABASE_URL"):
            raise RuntimeError(
                "[clear:db_postgres] DATABASE_URL not set. "
                "Add a Postgres database in your deploy target's dashboard."
            )
        # psycopg3's ConnectionPool — lazy connection, automatic recycling.
        # Match db-postgres.js's connection string pattern.
        from psycopg_pool import ConnectionPool  # type: ignore
        _pool = ConnectionPool(os.environ["DATABASE_URL"], min_size=0, max_size=10)
        return _pool


# In-memory schema registry (synchronous — matches db-postgres.js)
_schemas: Dict[str, Dict[str, Dict[str, Any]]] = {}
_tables_created: set = set()
_IDENT_RE = re.compile(r"^[a-z_][a-z0-9_]*$", re.IGNORECASE)


def _is_safe_identifier(name: str) -> bool:
    return bool(_IDENT_RE.match(name))


_PG_TYPE_MAP = {
    "number": "DOUBLE PRECISION",
    "boolean": "BOOLEAN",
    "fk": "INTEGER",
    "timestamp": "TIMESTAMPTZ",
}


def _to_pg_type(config: Optional[Dict[str, Any]]) -> str:
    if not config or not config.get("type"):
        return "TEXT"
    return _PG_TYPE_MAP.get(config["type"], "TEXT")


def _coerce_record(record: Optional[Dict[str, Any]], schema: Optional[Dict[str, Any]]) -> Optional[Dict[str, Any]]:
    """Postgres returns proper Python types via psycopg's adapters, so
    boolean coercion (which SQLite needs) is mostly a no-op here. Hooks
    are kept for parity with db.py and for future sensitive-field
    decryption integration."""
    if not record or not schema:
        return record
    return dict(record)


def _build_where(filter_dict: Optional[Dict[str, Any]], param_offset: int = 0):
    """Build a parameterized WHERE clause. Postgres uses $1, $2 syntax via
    psycopg's parameter style — we use %s which psycopg translates."""
    if not filter_dict:
        return {"sql": "", "params": []}
    keys = list(filter_dict.keys())
    for k in keys:
        if not _is_safe_identifier(k):
            raise ValueError(f"Unsafe column name: {k}")
    parts = [f"{k} = %s" for k in keys]
    params = [filter_dict[k] for k in keys]
    return {"sql": "WHERE " + " AND ".join(parts), "params": params}


def _ensure_table(table: str) -> None:
    """Lazy CREATE TABLE on first query against a registered schema.
    Matches db-postgres.js's lazy-init pattern — schemas are recorded
    synchronously by create_table(), tables are created on first use."""
    if table in _tables_created:
        return
    schema = _schemas.get(table)
    if not schema:
        return  # raw_query case — no schema registered
    full_schema = dict(schema)
    cols = ["id SERIAL PRIMARY KEY"]
    for field, config in full_schema.items():
        if not _is_safe_identifier(field):
            raise ValueError(f"Unsafe column name: {field}")
        if field == "id":
            continue
        col_def = f"{field} {_to_pg_type(config)}"
        if config.get("required"):
            col_def += " NOT NULL"
        if "default" in config:
            d = config["default"]
            if isinstance(d, str):
                col_def += f" DEFAULT '{d}'"
            elif isinstance(d, bool):
                col_def += f" DEFAULT {'TRUE' if d else 'FALSE'}"
            elif d is not None:
                col_def += f" DEFAULT {d}"
        cols.append(col_def)
    # Auto-add user_id / tenant_id / _version (mirrors db-postgres.js)
    if "user_id" not in full_schema:
        cols.append("user_id INTEGER")
    if "tenant_id" not in full_schema:
        cols.append("tenant_id INTEGER")
    if "_version" not in full_schema:
        cols.append("_version INTEGER DEFAULT 1")
    sql = f"CREATE TABLE IF NOT EXISTS {table} ({', '.join(cols)})"
    with _get_pool().connection() as conn:
        conn.execute(sql)
    _tables_created.add(table)


# ----------------------------------------------------------------------
# Public API — matches runtime/db.py
# ----------------------------------------------------------------------

def create_table(name: str, schema: Dict[str, Dict[str, Any]]) -> None:
    """Register schema. Actual CREATE TABLE happens lazily on first use
    via _ensure_table() because module load may happen before the pool
    is connectable (e.g. health-check-only servers)."""
    if not _is_safe_identifier(name):
        raise ValueError(f"Unsafe table name: {name}")
    _schemas[name] = schema


def find_all(table: str, filter_dict: Optional[Dict[str, Any]] = None,
             options: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:
    if not _is_safe_identifier(table):
        raise ValueError(f"Unsafe table name: {table}")
    _ensure_table(table)
    where = _build_where(filter_dict)
    sql = f"SELECT * FROM {table} {where['sql']}"
    options = options or {}
    limit = options.get("limit")
    offset = options.get("offset")
    if limit is not None:
        sql += f" LIMIT {int(limit)}"
    if offset is not None:
        sql += f" OFFSET {int(offset)}"
    schema = _schemas.get(table)
    with _get_pool().connection() as conn:
        with conn.cursor(row_factory=dict_row) as cur:
            cur.execute(sql, where["params"])
            rows = cur.fetchall()
    return [_coerce_record(r, schema) for r in rows]


def find_one(table: str, filter_dict: Dict[str, Any],
             options: Optional[Dict[str, Any]] = None) -> Optional[Dict[str, Any]]:
    results = find_all(table, filter_dict, {"limit": 1, **(options or {})})
    return results[0] if results else None


def insert(table: str, record: Dict[str, Any]) -> Dict[str, Any]:
    if not _is_safe_identifier(table):
        raise ValueError(f"Unsafe table name: {table}")
    _ensure_table(table)
    keys = [k for k in record.keys() if _is_safe_identifier(k)]
    placeholders = ", ".join(["%s"] * len(keys))
    cols = ", ".join(keys)
    params = [record[k] for k in keys]
    sql = f"INSERT INTO {table} ({cols}) VALUES ({placeholders}) RETURNING *"
    with _get_pool().connection() as conn:
        with conn.cursor(row_factory=dict_row) as cur:
            cur.execute(sql, params)
            row = cur.fetchone()
    return dict(row) if row else {**record}


def update(table: str, filter_or_record: Union[int, Dict[str, Any]],
           data: Optional[Dict[str, Any]] = None) -> int:
    """Two call shapes (matches db.py + db.js):
    - update(table, record_with_id) — uses record's id, updates all other fields
    - update(table, filter_dict, data_dict) — explicit WHERE + SET"""
    if not _is_safe_identifier(table):
        raise ValueError(f"Unsafe table name: {table}")
    _ensure_table(table)
    if data is None:
        if not isinstance(filter_or_record, dict) or "id" not in filter_or_record:
            raise ValueError("update(table, record) requires an `id` field on the record")
        rec = dict(filter_or_record)
        record_id = rec.pop("id")
        filter_dict = {"id": record_id}
        data = rec
    else:
        filter_dict = filter_or_record if isinstance(filter_or_record, dict) else {"id": filter_or_record}

    set_keys = [k for k in data.keys() if _is_safe_identifier(k)]
    set_clause = ", ".join([f"{k} = %s" for k in set_keys])
    set_params = [data[k] for k in set_keys]
    where = _build_where(filter_dict)
    sql = f"UPDATE {table} SET {set_clause} {where['sql']}"
    with _get_pool().connection() as conn:
        with conn.cursor() as cur:
            cur.execute(sql, set_params + where["params"])
            return cur.rowcount


def update_with_version(table: str, filter_dict: Dict[str, Any], data: Dict[str, Any]) -> int:
    """Optimistic-lock UPDATE. Stubbed for v1 — matches db.py's pattern."""
    raise NotImplementedError(
        "update_with_version is the optimistic-lock primitive — port deferred. "
        "Tracked in plans/plan-python-parity.md as priority follow-up."
    )


def remove(table: str, filter_dict: Optional[Dict[str, Any]] = None) -> int:
    if not _is_safe_identifier(table):
        raise ValueError(f"Unsafe table name: {table}")
    _ensure_table(table)
    where = _build_where(filter_dict)
    sql = f"DELETE FROM {table} {where['sql']}".strip() if where["sql"] else f"DELETE FROM {table}"
    with _get_pool().connection() as conn:
        with conn.cursor() as cur:
            cur.execute(sql, where["params"])
            return cur.rowcount


_AGG_FNS = {"sum": "SUM", "avg": "AVG", "min": "MIN", "max": "MAX", "count": "COUNT"}


def aggregate(table: str, fn: str, field: str, filter_dict: Optional[Dict[str, Any]] = None) -> Optional[Union[int, float]]:
    if not _is_safe_identifier(table) or not _is_safe_identifier(field):
        raise ValueError(f"Unsafe identifier: table={table}, field={field}")
    sql_fn = _AGG_FNS.get(fn.lower())
    if not sql_fn:
        raise ValueError(f"Unknown aggregate function: {fn}")
    _ensure_table(table)
    where = _build_where(filter_dict)
    sql = f"SELECT {sql_fn}({field}) AS result FROM {table} {where['sql']}"
    with _get_pool().connection() as conn:
        with conn.cursor(row_factory=dict_row) as cur:
            cur.execute(sql, where["params"])
            row = cur.fetchone()
    return row["result"] if row else None


def run(sql: str) -> None:
    with _get_pool().connection() as conn:
        conn.execute(sql)


def execute(sql: str) -> None:
    run(sql)


def save() -> None:
    """No-op. Postgres is always durable (matches db-postgres.js)."""


def load() -> None:
    """No-op. Pool initializes lazily on first query."""


def reset() -> None:
    """TRUNCATE all known tables. Used by test harnesses."""
    if not _tables_created:
        return
    names = ", ".join(t for t in _tables_created if _is_safe_identifier(t))
    if names:
        with _get_pool().connection() as conn:
            conn.execute(f"TRUNCATE {names} RESTART IDENTITY CASCADE")


def close() -> None:
    global _pool
    if _pool is not None:
        _pool.close()
        _pool = None


def _reset_for_tests() -> None:
    """Test helper: clears the schema registry. Doesn't drop tables."""
    _schemas.clear()
    _tables_created.clear()
